src/util/Tokenizer.ts

```diff
@@ -1,60 +1,225 @@
+import FileSystem from "./FileSystem";
+
 export class TokenizerError extends Error {
     constructor(...args: any[]) {
         super(...args);
         Error.captureStackTrace(this, TokenizerError);
     }
 }
 
+
 export class Tokenizer {
-    private static tokenizerInstance: Tokenizer;
-    private static tokenizerFile: string;
-    private tokens: string[];
+    private tokenizerFile: string;
+    private program: string;
+    private tokens: string[][];
+    private currentToken: {arr: number, pos: number};
+
+    public static literals: {[id: string] : {val: string, isSpecial: boolean}} =
+        Object.freeze({
+        Program: {val: "program", isSpecial: false},
+        Modules: {val: "modules", isSpecial: false},
+        Dir: {val: "dir", isSpecial: false},
+        Class: {val: "class", isSpecial: false},
+        Interface: {val: "interface", isSpecial: false},
+        Field: {val: "fields", isSpecial: false},
+        Generate: {val: "generates", isSpecial: false},
+        Function: {val: "function", isSpecial: false},
+        Async: {val: "async", isSpecial: false},
+        Static: {val: "static", isSpecial: false},
+        Returns: {val: "returns", isSpecial: false},
+        Params: {val: "params", isSpecial: false},
+        Comments: {val: "comments", isSpecial: false},
+        Abstract: {val: "abstract", isSpecial: false},
+        Implements: {val: "implements", isSpecial: false},
+        Extends: {val: "extends", isSpecial: false},
+        Comma: {val: "\,", isSpecial: true},
+        SquareBracketStart: {val: "\[", isSpecial: true},
+        SquareBracketEnd: {val: "\]", isSpecial: true},
+        Getters: {val: "getters", isSpecial: false},
+        Setters: {val: "setters", isSpecial: false},
+        Private: {val: "private", isSpecial: false},
+        Public: {val: "public", isSpecial: false},
+        Protected: {val: "protected", isSpecial: false},
+        SingleQuote: {val: "\'", isSpecial: true},
+        DoubleQuote: {val: "\"", isSpecial: true},
+        Constructor: {val: "constructor", isSpecial: false}
+    });
+    public static NO_MORE_TOKENS = "NO_MORE_TOKENS";
+    private static reservedTokenWord = "_RESERVED_";
+
 
-    private constructor(fileName: string) {
-        Tokenizer.tokenizerFile =  fileName;
-        // TODO: implement this
+    constructor(fileName: string, path: string) {
+        // initialize fields
+        this.tokenizerFile =  `${path}/${fileName}`;
```

Comment:
not needed?

Reply:
Thanks! removed this in latest commit (only change made in commit)